File: sdk/python/xibalba_integrity/scoring.py
"""
Xibalba Integrity SDK — Scoring Engine (Offline)

A portable copy of the Tri-Metric Scoring Engine that runs client-side.
Developers can compute scores locally before reporting to Xibalba for
hash generation and on-chain anchoring.
"""
import math
import statistics
from typing import List


class ScoringEngine:
    """Client-side scoring engine for the Integrity Protocol.

    Computes the three core metrics:
        1. Entropy Score  — measures performance stability
        2. Grounding Score — measures human oversight quality
        3. Integrity Score — the composite AIS (Agent Integrity Score)
    """

    MAX_SCORE = 1000

    # Composite Integrity Score weights (sum = 1.0)
    W_TRUSTFLOW    = 0.25   # Recursive partner reputation
    W_XIBALBA      = 0.25   # Xibalba audit verification
    W_SACRIFICE    = 0.20   # Verified compute hours
    W_STAKING_AGE  = 0.15   # Staking + longevity
    W_VOLUME       = 0.15   # Transaction volume

    # ── Metric 1: Entropy (Stability) ──────────────────────────────

    @staticmethod
    def entropy_score(performance_variance: float) -> int:
        """Calculate the Entropy Score from performance variance.

        A low variance yields a high score (stable agent).
        Formula: S = e^(-1.5 * variance^2) * 1000

        Args:
            performance_variance: Coefficient of variation across recent
                                  latency and accuracy measurements (0.0–2.0).

        Returns:
            Integer score 0–1000.
        """
        stability = math.exp(-1.5 * (performance_variance ** 2))
        return round(stability * ScoringEngine.MAX_SCORE)

    @staticmethod
    def calculate_variance(latencies: List[float], accuracies: List[float]) -> float:
        """Derive combined performance variance from raw telemetry.

        Uses Coefficient of Variation (CV) weighted 60/40 between
        latency and accuracy.

        Args:
            latencies: List of response latencies in ms.
            accuracies: List of accuracy scores (0.0 – 1.0).

        Returns:
            Combined variance float (0.0 – 2.0).
        """
        if len(latencies) < 2:
            return 0.5  # Insufficient data — assume moderate variance

        mean_lat = statistics.mean(latencies)
        mean_acc = statistics.mean(accuracies)

        cv_lat = statistics.stdev(latencies) / mean_lat if mean_lat > 0 else 1.0
        cv_acc = statistics.stdev(accuracies) / mean_acc if mean_acc > 0 else 1.0

        return round(min(2.0, (cv_lat * 0.6) + (cv_acc * 0.4)), 4)

    # ── Metric 2: Grounding (Accountability) ──────────────────────

    @staticmethod
    def grounding_score(hgi_raw: float) -> int:
        """Calculate the Grounding Score from the Human Grounding Index.

        Args:
            hgi_raw: Human Grounding Index value (0.0 – 1.0).

        Returns:
            Integer score 0–1000.
        """
        return round(hgi_raw * ScoringEngine.MAX_SCORE)

    # ── Metric 3: Integrity (Composite AIS) ───────────────────────

    def integrity_score(
        self,
        avg_partner_ais: float = 500,
        xibalba_audit_score: float = 0.0,
        gpu_hours_verified: float = 0,
        hgi_raw: float = 0.0,
        performance_variance: float = 0.5,
        staked_ratio: float = 0.0,
        agent_age_days: int = 1,
        total_volume_intg: float = 0,
        days_since_active: int = 0,
        penalty_points: float = 0.0,
    ) -> dict:
        """Calculate the full Tri-Metric Trust Profile.

        Returns a dict with entropy_score, grounding_score,
        integrity_score, stability_drag, and grounding_boost.
        """
        # Entropy
        e_score = self.entropy_score(performance_variance)
        stability_drag = e_score / self.MAX_SCORE

        # Grounding
        g_score = self.grounding_score(hgi_raw)
        grounding_boost = 1.0 + (hgi_raw * 0.2)

        # Component indices
        trustflow_idx   = min(1.0, avg_partner_ais / 1000.0)
        audit_idx       = min(1.0, max(0.0, xibalba_audit_score))
        sacrifice_idx   = min(1.0, math.log10(gpu_hours_verified + 1) / 3.0)
        age_idx         = min(1.0, math.log10(agent_age_days + 1) / 2.56)
        staking_age_idx = (0.5 * staked_ratio) + (0.5 * age_idx)
        volume_idx      = min(1.0, math.log10(total_volume_intg + 1) / 6.0)

        base_integrity = (
            (self.W_TRUSTFLOW   * trustflow_idx)  +
            (self.W_XIBALBA     * audit_idx)       +
            (self.W_SACRIFICE   * sacrifice_idx)   +
            (self.W_STAKING_AGE * staking_age_idx) +
            (self.W_VOLUME      * volume_idx)
        )

        correlated = base_integrity * stability_drag * grounding_boost

        # Penalties & decay
        penalty_mul  = 1.0 - min(1.0, penalty_points)
        temporal_dec = math.exp(-0.005 * days_since_active)

        final = min(self.MAX_SCORE, correlated * self.MAX_SCORE * penalty_mul * temporal_dec)

        return {
            "entropy_score":   e_score,
            "grounding_score": g_score,
            "integrity_score": round(final),
            "stability_drag":  round(stability_drag, 4),
            "grounding_boost": round(grounding_boost, 4),
        }

File: sdk/nodejs/src/scoring.ts
/**
 * Xibalba Integrity SDK — Client-Side Scoring Engine
 *
 * Portable TypeScript implementation of the Tri-Metric Scoring Engine.
 * Developers can compute scores locally before reporting to Xibalba.
 *
 * @module @xibalba/integrity-sdk/scoring
 */

import type { TriMetricResult, IntegrityScoreParams } from "./types";

const MAX_SCORE = 1000;

// Composite Integrity Score weights (sum = 1.0)
const W_TRUSTFLOW = 0.25;
const W_XIBALBA = 0.25;
const W_SACRIFICE = 0.20;
const W_STAKING_AGE = 0.15;
const W_VOLUME = 0.15;

export class ScoringEngine {
  /**
   * Metric 1: Entropy Score (Stability).
   * A low variance yields a high score — the agent is predictable and reliable.
   *
   * @param performanceVariance - Combined CV of latency + accuracy (0.0–2.0).
   * @returns Integer score 0–1000.
   */
  static entropyScore(performanceVariance: number): number {
    const stability = Math.exp(-1.5 * (performanceVariance ** 2));
    return Math.round(stability * MAX_SCORE);
  }

  /**
   * Derive combined performance variance from raw telemetry arrays.
   *
   * @param latencies - Array of response latencies in ms.
   * @param accuracies - Array of accuracy scores (0.0–1.0).
   * @returns Combined variance (0.0–2.0).
   */
  static calculateVariance(latencies: number[], accuracies: number[]): number {
    if (latencies.length < 2) return 0.5;

    const meanLat = latencies.reduce((a, b) => a + b, 0) / latencies.length;
    const meanAcc = accuracies.reduce((a, b) => a + b, 0) / accuracies.length;

    const stdLat = Math.sqrt(
      latencies.reduce((sum, v) => sum + (v - meanLat) ** 2, 0) / (latencies.length - 1)
    );
    const stdAcc = Math.sqrt(
      accuracies.reduce((sum, v) => sum + (v - meanAcc) ** 2, 0) / (accuracies.length - 1)
    );

    const cvLat = meanLat > 0 ? stdLat / meanLat : 1.0;
    const cvAcc = meanAcc > 0 ? stdAcc / meanAcc : 1.0;

    return Math.min(2.0, parseFloat((cvLat * 0.6 + cvAcc * 0.4).toFixed(4)));
  }

  /**
   * Metric 2: Grounding Score (Accountability).
   *
   * @param hgiRaw - Human Grounding Index value (0.0–1.0).
   * @returns Integer score 0–1000.
   */
  static groundingScore(hgiRaw: number): number {
    return Math.round(hgiRaw * MAX_SCORE);
  }

  /**
   * Calculate the full Tri-Metric Trust Profile.
   *
   * @param params - All scoring parameters.
   * @returns TriMetricResult with all three scores and correlation factors.
   */
  static integrityScore(params: IntegrityScoreParams = {}): TriMetricResult {
    const {
      avgPartnerAis = 500,
      xibalbaAuditScore = 0.0,
      gpuHoursVerified = 0,
      hgiRaw = 0.0,
      performanceVariance = 0.5,
      stakedRatio = 0.0,
      agentAgeDays = 1,
      totalVolumeIntg = 0,
      daysSinceActive = 0,
      penaltyPoints = 0.0,
      verificationTier = 1,
    } = params;

    // Entropy
    const eScore = ScoringEngine.entropyScore(performanceVariance);
    const stabilityDrag = eScore / MAX_SCORE;

    // Grounding
    const gScore = ScoringEngine.groundingScore(hgiRaw);
    const groundingBoost = 1.0 + hgiRaw * 0.2;

    // Component indices
    const trustflowIdx = Math.min(1.0, avgPartnerAis / 1000.0);
    const auditIdx = Math.min(1.0, Math.max(0.0, xibalbaAuditScore));
    const sacrificeIdx = Math.min(1.0, Math.log10(gpuHoursVerified + 1) / 3.0);
    const ageIdx = Math.min(1.0, Math.log10(agentAgeDays + 1) / 2.56);
    const stakingAgeIdx = 0.5 * stakedRatio + 0.5 * ageIdx;
    const volumeIdx = Math.min(1.0, Math.log10(totalVolumeIntg + 1) / 6.0);

    const baseIntegrity =
      W_TRUSTFLOW * trustflowIdx +
      W_XIBALBA * auditIdx +
      W_SACRIFICE * sacrificeIdx +
      W_STAKING_AGE * stakingAgeIdx +
      W_VOLUME * volumeIdx;

    const correlated = baseIntegrity * stabilityDrag * groundingBoost;

    // Penalties & decay
    const penaltyMul = 1.0 - Math.min(1.0, penaltyPoints);
    const temporalDec = Math.exp(-0.005 * daysSinceActive);

    let final = correlated * MAX_SCORE * penaltyMul * temporalDec;

    // ENFORCE IDENTITY CEILING
    let ceiling = 600; // Tier 1 Default
    if (verificationTier === 2) {
      ceiling = 850;
    } else if (verificationTier === 3) {
      ceiling = 1000;
    }
    final = Math.min(final, ceiling);

    return {
      entropyScore: eScore,
      groundingScore: gScore,
      integrityScore: Math.round(Math.max(0, final)),
      stabilityDrag: parseFloat(stabilityDrag.toFixed(4)),
      groundingBoost: parseFloat(groundingBoost.toFixed(4)),
    };
  }
}
